### PR TITLE
ctexplain: make cquery parsing more robust

### PR DESCRIPTION
**Pull request by gregestren**
Specifically:

 - Handle labels with spaces
 - Handle targets with no fragment requirements.

These were both inspired by real analyses.

Testing: `$ bazel test //tools/ctexplain:all`
Work towards #10613

### CODE REVIEW

File: tools/ctexplain/bazel_api.py [hunk 1]
```diff
@@ -145,20 +145,27 @@ def _parse_cquery_result_line(line: str) -> ConfiguredTarget:
...
+    if fragments_start == -1 or fragments_end < fragments_start:
+      raise ValueError(f'"{line}": no "[<required fragment>, ...]" suffix')
     # The fragments list looks like '[Fragment1, Fragment2, ...]'. Split the
     # whole line on ' [' to get just this list, then remove the final ']', then
     # split again on ', ' to convert it to a structured tuple.
-    fragments = tuple(line.split(" [")[1][0:-1].split(", "))
+    if fragments_start == fragments_end - 1:
+      fragments = tuple()
+    else:
+      fragments = tuple(line[fragments_start + 1:fragments_end].split(", "))
```
Comment (sdtwigg) [inline tools/ctexplain/bazel_api.py:168]: tuple(x.strip() for x in line[fragments_start + 1:fragments_end].split(","))

is a little more resilient to whitespace changes.

File: tools/ctexplain/bazel_api.py [hunk 2]
```diff
@@ -145,20 +145,27 @@ def _parse_cquery_result_line(line: str) -> ConfiguredTarget:
...
     Corresponding ConfiguredTarget if the line matches else None.
   """
-  tokens = line.split(maxsplit=2)
-  label = tokens[0]
-  if tokens[1][0] != "(" or tokens[1][-1] != ")":
-    raise ValueError(f"{tokens[1]} in {line} not surrounded by parentheses")
-  config_hash = tokens[1][1:-1]
+  config_hash_start = line.find("(")
+  config_hash_end = line.find(")", config_hash_start)
+  fragments_start = line.find("[", config_hash_end)
```
Comment (sdtwigg) [inline tools/ctexplain/bazel_api.py:150]: move under the else: condition below; after verifying not in null configuration

File: tools/ctexplain/bazel_api_test.py [hunk 3]
```diff
@@ -144,14 +144,38 @@ def testConfigWithStarlarkFlags(self):
...
+  def testLabelsWithSpaces(self):
+    self.ScratchFile('testapp/BUILD', [
+        'filegroup(name = "labels can have spaces", srcs = ["a.file"])',
+    ])
+    cquery_args = ['//testapp:labels can have spaces']
+    res = self._bazel_api.cquery(['//testapp:all'])
+    success = res[0]
+    cts = res[2]
+    self.assertTrue(success)
+    self.assertEqual(len(cts), 1)
```
Comment (sdtwigg) [inline tools/ctexplain/bazel_api_test.py:164]: assertLen

File: tools/ctexplain/bazel_api_test.py [hunk 4]
```diff
@@ -144,14 +144,38 @@ def testConfigWithStarlarkFlags(self):
...
+  def testEmptyFragmentRequirements(self):
+    self.ScratchFile('testapp/BUILD', [
+        'alias(name = "alias_to_src", actual = "source.file")',
+    ])
+    cquery_args = ['//testapp:alias_to_src']
+    res = self._bazel_api.cquery(['//testapp:all'])
+    success = res[0]
+    cts = res[2]
+    self.assertTrue(success)
+    self.assertEqual(len(cts), 1)
```
Comment (sdtwigg) [inline tools/ctexplain/bazel_api_test.py:176]: assertLen

File: tools/ctexplain/bazel_api_test.py [hunk 5]
```diff
@@ -144,14 +144,38 @@ def testConfigWithStarlarkFlags(self):
...
+    self.assertEqual(cts[0].label, '//testapp:alias_to_src')
+    self.assertEqual(len(cts[0].transitive_fragments), 0)
```
Comment (sdtwigg) [inline tools/ctexplain/bazel_api_test.py:178]: assertEmpty

**Review by sdtwigg — CHANGES_REQUESTED** [review]

**Comment by keertk** [comment]
Hi all, we're doing a clean up of old PRs and will be closing this one since it seems to have stalled. Please feel free to reopen if you’re still interested in pursuing this.

**Pull request closed without merging**